### PR TITLE
KEYCLOAK-5571 Changes to support emailVerified boolean mapper with LD…

### PR DESCRIPTION
**Pull request by 1rahulr**
KEYCLOAK-5571 Changes to support emailVerified boolean mapper with LDAP user federation.

LDAP Boolean attributes accept "TRUE" or "FALSE" (https://tools.ietf.org/html/rfc4517#page-7).
java Boolean to LDAP Boolean mapping done in case mapper is of LDAP Boolean type.

### CODE REVIEW

**Comment by stianst** [comment]
Looks like a different approach to https://github.com/keycloak/keycloak/pull/6160

**Comment by stianst** [comment]
6610 looks like the best approach to me

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java [hunk 1]
```diff
@@ -453,14 +471,23 @@ private boolean isReadOnly() {
...
+            }	
             userModelProperty.setValue(user, null);
         } else {
             Class<Object> clazz = userModelProperty.getJavaClass();
 
             if (String.class.equals(clazz)) {
                 userModelProperty.setValue(user, ldapAttrValue);
             } else if (Boolean.class.equals(clazz) || boolean.class.equals(clazz)) {
-                Boolean boolVal = Boolean.valueOf(ldapAttrValue);
+   		Boolean boolVal = false;
```
Comment (mposolda) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java]: Should be better to keep "Boolean boolVal = Boolean.valueOf(ldapAttrValue)" . That should handle the case when ldapAttrValue is "TRUE" or "true" (probably better regarding compatibility among various LDAP servers)

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java [hunk 2]
```diff
@@ -259,6 +269,12 @@ public void setFirstName(String firstName) {
                     super.setFirstName(firstName);
                 }
 
+                @Override
+                public void setEmailVerified(boolean verified) {
```
Comment (mposolda) [inline federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java]: I am seeing "setEmailVerified" overriden, but not "isEmailVerified" . Is this a mistake?

**Review by mposolda — CHANGES_REQUESTED** [review]
@1rahulr Thanks, but I am not sure if this PR can be accepted. Few questions:

- Is it some standard LDAP attribute to which can be "emailVerified" property mapped? If yes, could you point to particular RFC or something?
- Formatting is a bit broken. We use 4 spaces per line, no tabs. See the existing codebase for details and configure your IDE properly. I've also added 2 minor comments, but I would like to have answer to the previous question before evaluate if PR can be accepted or not.

Thanks

**Comment by 1rahulr** [comment]
HI,
emailVerified was added as custom user attribute 

ldapmodify -v -D "${ADMIN}" -w "${PASSWORD}" <<!
dn: cn=schema
changetype: modify
add: attributeTypes
attributeTypes: ( 2.25.28639311321113238241701611583088740684.14.2.1
  NAME 'emailVerified'
  EQUALITY booleanMatch
  SYNTAX 1.3.6.1.4.1.1466.115.121.1.7 )
-
add: objectClasses
objectClasses: ( 2.25.28639311321113238241701611583088740684.14.2.2
  NAME 'emailVerification'
  DESC 'emailVerification'
  SUP inetOrgPerson
  STRUCTURAL
  MAY (emailVerified)
)
!

**Comment by mposolda** [comment]
@1rahulr Thanks for pointing LDAP attribute and sorry for late response. How about remaining comments mentioned in this PR? Also I think that this may need some documentation with the example of extending LDAP schema to support boolean attributes like "emailVerified" /

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java (hunk 2 above)
Comment (1rahulr) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java]: No need to override "isEmailVerified". Even without this pull request read of emailVerified attribute is working fine, only write to LDAP had issues.

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java (hunk 1 above)
Comment (1rahulr) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java]: Agree. Updated it.

**Comment by 1rahulr** [comment]
@mposolda : Sorry for the delay form my part. I have tried to address the comments and updated the formatting. Please check. Any suggestions for where the documentation should be updated ? Is a new using-emailVerified-with-ldap.md file in docs folder sufficient ?

File: federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java (hunk 2 above)
Comment (mposolda) [reply federation/ldap/src/main/java/org/keycloak/storage/ldap/mappers/UserAttributeLDAPStorageMapper.java]: AFAIK if you don't override "isEmailVerified", then it reads the value of emailVerified from the Keycloak DB. Which can be fine if it has same value as LDAP. However for example in the case like this:
- Keycloak user is registered with emailVerified "true"
- Later administrator directly update user in LDAP and set "emailVerified" to false
- Keycloak will now incorrectly read emailVerified "true" from the Keycloak DB without considering newly updated value from the LDAP

**Review by mposolda — CHANGES_REQUESTED** [review]
@1rahulr Thanks and sorry again for the late update. Added just one inline comment regarding "isEmailVerified" .

Besides that, the documentation should be updated in the KEycloak documentation under https://github.com/keycloak/keycloak-documentation github repository. You can add some new section into the `LDAP` chapter in the server admin guide. The location under "docs/" in the keycloak repository is not good, so feel free to remove it from this PR.

For the docs, my suggestion would be to remove the need of "emailVerification" to be added as required objectClass. Also I will remove all the IPA specific commands for update existing users (everything after line 37 in the document). Thanks

**Comment by mposolda** [comment]
Adding label "Hold" as this is not a blocker for Keycloak 9.0

**Comment by mposolda** [comment]
@1rahulr Any chance to continue on this PR and address the remaining concerns? Also the documentation PR will be good.

**Comment by mposolda** [comment]
@1rahulr Closing this as remaining concerns are not addressed and documentation is not made. Feel free to re-open if you address remaining concerns. Since this change is not a blocker and requires custom changes to LDAP schema, I don't insist on this feature be widely available.

**Pull request closed without merging**